content/en/blog/attribute-autocomplete-enhancement/index.md

```diff
@@ -0,0 +1,306 @@
+
+**![](https://lh3.googleusercontent.com/vqgcIM-ae7mRCIbah-ZyZRJkM4T4MMNdo3HGQtVT0RFfeh9J69z7SIUspMfY1IUjJ0V30U84PLdBMd3QouXKLEbfjUBc3Ac5q3_CtKk5XAP55I3QY0oG2U6Q11TIWcXUrVfjhbGpUUgV7LPp1R4ACeM)**
+
+Fig. 2, Dropdown menu for attribute autocomplete
+
+## Auto Attribute Correction
+
+### Motivation
+
+In the previous autocomplete implementation, when an attribute name in the input schema of an operator changes, the field of this operator that selects the attribute will not update its value according to the change in the attribute name, and the operator will therefore become invalid. Intuitively, it is expected that the field will be updated accordingly. 
+
+Therefore, we introduce the auto attribute correction feature that automatically corrects the attribute selection in operators whenever an attribute is deleted or renamed.
+
+### Demonstration
+
+The auto attribute correction feature updates the attribute selection in an operator under the following circumstances
+
+1. when the name of the attribute selected in the operator changes,
+2. when the attribute selected in the operator is deleted
+
+#### When attribute is renamed
+
+When user renames an attribute through an operator, the attribute name will be updated in all succeeding operators.
+
+Fig. 3 shows how the attribute selection is automatically updated in the "Filter" operator when the result attribute name is changed in its preceding "Sentiment Analysis" operator.
+
+When the user changes the name of the output attribute of the "Sentiment Analysis" operator from `"A"` to `"B"`, the `Attribute` field whose value was `"A"` in the succeeding “Filter” operator is automatically updated to `"B"`.
+
+**![](https://lh6.googleusercontent.com/Zf6wNg1uX9m_zJLkqtcKAap8kNLQttz1whOKPG1xGFolZ6it0hYMIMw2hzCLukabM0P2hoJ43M4QIjIcULYP_mAKt0N3V82L-0yuK0FOLD9xjrXSx0DSSzlqZD9Qp8QUL1KvABBpbbw5pXXrGgVpqQ)**
+
+Fig. 3, Auto correction process when attribute is renamed
+
+#### When attribute is deleted
+
+In case of an attribute selected in an operator is deleted, e.g., deselected in a projection operator, all succeeding operators containing the attribute will delete the attribute from themselves and become invalid.
+
+### Implementation
+
+To automatically correct the attribute selection in operators' properties, we need to do the following steps:
+
+1. Trigger auto correction the input attributes, i.e. input schema, changes
+2. Determine what input attribute changes (rename or delete)
+3. Correct the attribute selection
+
+The first step is done by listening to the input schema change event. We will focus on the implementation of the second and third steps.
+
+#### Determine attribute change
+
+An operator knows an attribute changes when the input schema changes. However, it can be challenging to determine what attribute is changed and what it changes to because all an operator knows about its input is the input schema, which is a list of attribute names and types.
```

Comment:
`An operator knows an attribute changes when the input schema changes.` Unclear. what do you mean by `attribute change`? what do you mean by `knows`?

Reply:
Sorry, it was confusing. I changed it to "It is easy for an operator to know when its input schema is changed. However, ..."